esptool.py

```diff
@@ -3075,6 +3115,92 @@ def make_image(args):
     image.save(args.output)
 
 
+def merge_image(args):
+    import espsecure
+
+    # Helper classes
+    class Bin(object):
+        def __init__(self, addr, file):
+            self.addr = addr
+            self.file = file
+            self.file_name = os.path.basename(file.name)
+            self.size = self.get_size()
+
+        def get_size(self):
+            self.file.seek(0, 2)  # seek to end
+            size = self.file.tell()
+            self.file.seek(0)
+            return size
+
+    class MultiBin(object):
+        def __init__(self, name, output_folder='.'):
+            self.name = name
+            self.output_folder = output_folder
+            try:
+                os.makedirs(os.path.realpath(self.output_folder))
+            except OSError:
+                pass    # file already exists, overwrite
+            self.output_path = os.path.realpath(os.path.join(self.output_folder, self.name))
+            self.bin_array = []
+
+        def add_bin(self, addr, file):
+            self.bin_array.append(Bin(addr, file))
+
+        def sort_bin(self):
+            swapped = True
+            while swapped:
+                swapped = False
+                for i in range(len(self.bin_array) - 1):
+                    if self.bin_array[i].addr > self.bin_array[i + 1].addr:
+                        self.bin_array[i], self.bin_array[i + 1] = self.bin_array[i + 1], self.bin_array[i]
+                        swapped = True
+
+        def add_bin_to_other_bin(self, previous, binary):
+            with open(self.output_path, "a") as output_file:
+                output_file.write('\xff' * (binary.addr - previous))
+            print("Add %s from 0x%x to 0x%x (0x%x)" % (
+                binary.file_name, binary.addr, binary.addr + binary.size, binary.size))
+            with open(self.output_path, "a") as output_file:
+                output_file.write(binary.file.read())
+            return binary.addr + binary.size
+
+        def create_bin(self):
+            new_start = 0
+            open(self.output_path, "wb").close()
+            for b in self.bin_array:
+                new_start = self.add_bin_to_other_bin(new_start, b)
+
+        def check_if_possible(self):
+            for i in range(1, len(self.bin_array)):
+                if self.bin_array[i].addr <= (self.bin_array[i - 1].addr + self.bin_array[i - 1].size):
+                    print(self.bin_array[i].addr, (self.bin_array[i - 1].addr + self.bin_array[i - 1].size))
```

Comment:
debug statement left in by mistake?

It should be possible to use the `AddrFilenamePairAction` which is already used by `write_flash` command to test none of the binaries overlap.

Reply:
Removed debug statement. I've also cleaned up overlap check. But by no means if we have this functionality else where we should use it rather than rewrite it.